In this notebook, I will explore how to make the model more resilient to attacks.

**Task: Adversarial attacks on the model, implementation of
defensive mechanisms**
1. Modify the model or the training dataset to make it
more resistant to attacks (for example, by using adversarial
training, normalizing the input data, or using dropout).
2. Test the protected model on the same adversarial
examples.

In [ ]:
pip install ultralytics adversarial-robustness-toolbox

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 32.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 57.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 31.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 50.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 87.0 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
   

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import transforms
from torchvision.datasets import SVHN
from torchvision.models import resnet18
from torch.utils.data import DataLoader

import numpy as np
from tqdm import tqdm
from art.estimators.classification import PyTorchClassifier
from art.attacks.evasion import FastGradientMethod, ProjectedGradientDescent
import torch.nn.functional as F
from torch.utils.data import Subset
import random

Load data:

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

transform = transforms.Compose([
    transforms.ToTensor()
])

train_dataset = SVHN(root="./data", split='train', download=True, transform=transform)
test_dataset = SVHN(root="./data", split='test', download=True, transform=transform)

train_loader = DataLoader(train_dataset, batch_size=256, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=256, shuffle=False)

# Подготовка данных для ART
X_test = torch.stack([img for img, _ in test_dataset]).numpy()
y_test = np.array([label for _, label in test_dataset])
y_test[y_test == 10] = 0

100%|██████████| 182M/182M [00:06<00:00, 26.6MB/s]
100%|██████████| 64.3M/64.3M [00:04<00:00, 14.0MB/s]


Creating models and an ART classifier. I use Resnet18, also YOLO can be usable here too if you add an adapter described in a previous notebook - `Create_adversarial_clean.ipynb`.

In [ ]:
model = resnet18(weights=None)
model.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
model.maxpool = nn.Identity()
model.fc = nn.Linear(model.fc.in_features, 10)
model = model.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-3)

classifier = PyTorchClassifier(
    model=model,
    loss=criterion,
    optimizer=optimizer,
    input_shape=(3, 32, 32),
    nb_classes=10,
    clip_values=(0.0, 1.0),
    device_type="gpu" if torch.cuda.is_available() else "cpu"
)


Training a model with adversarial training (FGSM):

In [ ]:
adv_trainer = FastGradientMethod(estimator=classifier, eps=0.03)

epochs = 5
for epoch in range(epochs):
    model.train()
    total_loss = 0
    print(f"\nEpoch {epoch + 1}/{epochs}")
    progress_bar = tqdm(train_loader, desc="Adversarial Training", leave=False)

    for images, labels in progress_bar:
        images, labels = images.to(device), labels.to(device)

        # Сгенерировать adversarial примеры (на CPU)
        model.eval()
        adv_images = adv_trainer.generate(x=images.detach().cpu().numpy())
        adv_images = torch.tensor(adv_images).to(device).float()
        model.train()

        # Объединяем clean + adversarial
        mixed_images = torch.cat([images, adv_images])
        mixed_labels = torch.cat([labels, labels])

        optimizer.zero_grad()
        output = model(mixed_images)
        loss = criterion(output, mixed_labels)
        loss.backward()
        optimizer.step()

        total_loss += loss.item()
        progress_bar.set_postfix(loss=loss.item())

    print(f"Epoch {epoch + 1} complete — Avg loss: {total_loss / len(train_loader):.4f}")



Epoch 1/5


Epoch 1 complete — Avg loss: 1.0136

Epoch 2/5


Epoch 2 complete — Avg loss: 0.6646

Epoch 3/5


Epoch 3 complete — Avg loss: 0.4617

Epoch 4/5


Epoch 4 complete — Avg loss: 0.3210

Epoch 5/5


Epoch 5 complete — Avg loss: 0.1866


Conducting an attack and checking accuracy:

In [ ]:
# 1. Переоборачиваем модель после adversarial training
classifier = PyTorchClassifier(
    model=model,
    loss=criterion,
    optimizer=optimizer,
    input_shape=(3, 32, 32),
    nb_classes=10,
    clip_values=(0.0, 1.0),
    device_type="gpu" if torch.cuda.is_available() else "cpu"
)

X_test_np = X_test.astype(np.float32)
y_test_np = y_test.astype(np.int64)

preds_clean = classifier.predict(X_test_np)
acc_clean = np.mean(np.argmax(preds_clean, axis=1) == y_test_np)
print(f"\n🎯 Accuracy on clean test set: {acc_clean * 100:.2f}%")

# 4. FGSM атака
fgsm_attack = FastGradientMethod(estimator=classifier, eps=0.03)
X_test_adv_fgsm = fgsm_attack.generate(x=X_test_np)

preds_fgsm = classifier.predict(X_test_adv_fgsm)
acc_fgsm = np.mean(np.argmax(preds_fgsm, axis=1) == y_test_np)
print(f"⚠️ Accuracy under FGSM attack: {acc_fgsm * 100:.2f}%")

# 5. PGD атака
pgd_attack = ProjectedGradientDescent(estimator=classifier, eps=0.03, eps_step=0.005, max_iter=40)
X_test_adv_pgd = pgd_attack.generate(x=X_test_np)

preds_pgd = classifier.predict(X_test_adv_pgd)
acc_pgd = np.mean(np.argmax(preds_pgd, axis=1) == y_test_np)
print(f"🔥 Accuracy under PGD attack: {acc_pgd * 100:.2f}%")



🎯 Accuracy on clean test set: 95.18%
⚠️ Accuracy under FGSM attack: 93.37%


PGD - Batches:   0%|          | 0/814 [00:00<?, ?it/s]

🔥 Accuracy under PGD attack: 6.59%


It is clear that the model is not resistant to PGD attacks. Let's work on this.

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = resnet18(pretrained=False)
model.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
model.maxpool = nn.Identity()
model.fc = nn.Linear(model.fc.in_features, 10)
model = model.to(device)

# Функции потерь и оптимизатор
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-3)

# Оборачиваем в ART
classifier = PyTorchClassifier(
    model=model,
    loss=criterion,
    optimizer=optimizer,
    input_shape=(3, 32, 32),
    nb_classes=10,
    clip_values=(0.0, 1.0),
)

# Настройка PGD-атаки
pgd_attack = ProjectedGradientDescent(
    estimator=classifier,
    norm=np.inf,
    eps=0.03,
    eps_step=0.005,
    max_iter=10,
    targeted=False,
    num_random_init=1,
    batch_size=32,
    verbose=False
)

# Adversarial training
epochs = 5
for epoch in range(epochs):
    model.train()
    total_loss = 0
    print(f"\n🌊 Epoch {epoch + 1}/{epochs}")
    progress_bar = tqdm(train_loader, desc="PGD Adversarial Training", leave=False)

    for images, labels in progress_bar:
        images, labels = images.to(device), labels.to(device)

        # Генерация adversarial примеров (на CPU для совместимости с ART)
        model.eval()
        adv_images = pgd_attack.generate(x=images.detach().cpu().numpy())
        adv_images = torch.tensor(adv_images).to(device).float()
        model.train()

        # Объединение clean + adversarial
        mixed_images = torch.cat([images, adv_images])
        mixed_labels = torch.cat([labels, labels])

        optimizer.zero_grad()
        output = model(mixed_images)
        loss = criterion(output, mixed_labels)
        loss.backward()
        optimizer.step()

        total_loss += loss.item()
        progress_bar.set_postfix(loss=loss.item())

    print(f"✅ Epoch {epoch + 1} complete — Avg loss: {total_loss / len(train_loader):.4f}")


/usr/local/lib/python3.11/dist-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)



🌊 Epoch 1/5


✅ Epoch 1 complete — Avg loss: 1.1245

🌊 Epoch 2/5


✅ Epoch 2 complete — Avg loss: 0.8045

🌊 Epoch 3/5


✅ Epoch 3 complete — Avg loss: 0.7393

🌊 Epoch 4/5


✅ Epoch 4 complete — Avg loss: 0.7016

🌊 Epoch 5/5


✅ Epoch 5 complete — Avg loss: 0.6676


In [ ]:
# Преобразуем тестовый датасет в numpy (если еще не сделали)
X_test_np = np.array([x.numpy() for x, _ in test_dataset])
y_test_np = np.array([y for _, y in test_dataset])

X_test_np = X_test_np.transpose((0, 2, 3, 1))  # [N, H, W, C]
X_test_np = X_test_np.transpose((0, 3, 1, 2))  # [N, C, H, W]


In [ ]:
# FGSM
fgsm = FastGradientMethod(estimator=classifier, eps=0.03)
X_test_adv_fgsm = fgsm.generate(x=X_test_np)

# PGD
pgd = ProjectedGradientDescent(estimator=classifier, eps=0.03, eps_step=0.005, max_iter=10)
X_test_adv_pgd = pgd.generate(x=X_test_np)


PGD - Batches:   0%|          | 0/814 [00:00<?, ?it/s]

In [ ]:
# Точность на clean
pred_clean = np.argmax(classifier.predict(X_test_np), axis=1)
acc_clean = np.mean(pred_clean == y_test_np) * 100

# Точность на FGSM
pred_fgsm = np.argmax(classifier.predict(X_test_adv_fgsm), axis=1)
acc_fgsm = np.mean(pred_fgsm == y_test_np) * 100

# Точность на PGD
pred_pgd = np.argmax(classifier.predict(X_test_adv_pgd), axis=1)
acc_pgd = np.mean(pred_pgd == y_test_np) * 100

print(f"🎯 Accuracy on clean test set: {acc_clean:.2f}%")
print(f"⚠️ Accuracy under FGSM attack: {acc_fgsm:.2f}%")
print(f"🔥 Accuracy under PGD attack: {acc_pgd:.2f}%")


🎯 Accuracy on clean test set: 92.85%
⚠️ Accuracy under FGSM attack: 63.01%
🔥 Accuracy under PGD attack: 54.61%


Great, after training, the model became much more resistant to PGD attacks, but the quality of FGSM attacks decreased. This means that we need to train on a dataset that includes images of both attacks if we want to protect ourselves from them.

In [ ]:
transform = transforms.Compose([
    transforms.ToTensor()
])

svhn_train = SVHN(root="./data", split='train', download=True, transform=transform)

# Преобразуем SVHN в тензоры
X_train_tensor = torch.stack([img for img, _ in svhn_train])
y_train_tensor = torch.tensor([label if label != 10 else 0 for _, label in svhn_train])  # исправляем метки

y_train_tensor = y_train_tensor.long()
y_train_tensor = y_train_tensor.long()



In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

random_transform = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
])

class AugmentedTensorDataset(torch.utils.data.Dataset):
    def __init__(self, tensors, transform=None):
        self.tensors = tensors
        self.transform = transform

    def __getitem__(self, index):
        x, y = self.tensors[0][index], self.tensors[1][index]
        img = transforms.ToPILImage()(x)
        if self.transform:
            img = self.transform(img)
        x = transforms.ToTensor()(img)
        return x, y

    def __len__(self):
        return len(self.tensors[0])

indices = random.sample(range(len(train_dataset)), 10000)  # Например, 10k примеров вместо 73k
train_dataset = Subset(train_dataset, indices)
train_loader = DataLoader(train_dataset, batch_size=128, shuffle=True)

In [ ]:
class ResNetWithDropout(nn.Module):
    def __init__(self, base_model, dropout_rate=0.3):
        super().__init__()
        self.features = nn.Sequential(*list(base_model.children())[:-1])
        self.dropout = nn.Dropout(dropout_rate)
        self.fc = nn.Linear(base_model.fc.in_features, 10)

    def forward(self, x):
        x = self.features(x)
        x = x.view(x.size(0), -1)
        x = self.dropout(x)
        x = self.fc(x)
        return x

base_model = resnet18(weights=None)
base_model.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
base_model.maxpool = nn.Identity()
model = ResNetWithDropout(base_model).to(device)

In [ ]:
class LabelSmoothingCrossEntropy(nn.Module):
    def __init__(self, smoothing=0.1):
        super().__init__()
        self.smoothing = smoothing
        self.confidence = 1.0 - smoothing

    def forward(self, pred, target):
        log_probs = F.log_softmax(pred, dim=-1)
        true_dist = torch.zeros_like(log_probs)
        true_dist.fill_(self.smoothing / (pred.size(1) - 1))
        true_dist.scatter_(1, target.unsqueeze(1), self.confidence)
        return torch.mean(torch.sum(-true_dist * log_probs, dim=-1))


In [ ]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-3)

classifier = PyTorchClassifier(
    model=model,
    loss=criterion,
    optimizer=optimizer,
    input_shape=(3, 32, 32),
    nb_classes=10,
    clip_values=(0.0, 1.0),
    device_type='gpu' if torch.cuda.is_available() else 'cpu'
)

fgsm = FastGradientMethod(estimator=classifier, eps=0.03)
pgd = ProjectedGradientDescent(estimator=classifier, eps=0.03, max_iter=10)

# Обучение на смеси атак ===
epochs = 5
for epoch in range(epochs):
    model.train()
    total_loss = 0
    print(f"\n🌪 Epoch {epoch + 1}/{epochs}")
    progress_bar = tqdm(train_loader, desc="Mixed Adversarial Training")

    for images, labels in progress_bar:
        images, labels = images.to(device), labels.to(device)

        model.eval()
        fgsm_adv = fgsm.generate(x=images.detach().cpu().numpy())
        pgd_adv = pgd.generate(x=images.detach().cpu().numpy())

        fgsm_adv = torch.tensor(fgsm_adv).to(device).float()
        pgd_adv = torch.tensor(pgd_adv).to(device).float()
        model.train()

        # Объединяем clean + FGSM + PGD
        mixed_images = torch.cat([images, fgsm_adv, pgd_adv])
        mixed_labels = torch.cat([labels, labels, labels])

        optimizer.zero_grad()
        outputs = model(mixed_images)
        loss = criterion(outputs, mixed_labels)
        loss.backward()
        optimizer.step()

        total_loss += loss.item()
        progress_bar.set_postfix(loss=loss.item())

    print(f"✅ Epoch {epoch + 1} complete — Avg loss: {total_loss / len(train_loader):.4f}")


🌪 Epoch 1/5


Mixed Adversarial Training:   0%|          | 0/79 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   1%|▏         | 1/79 [00:01<02:34,  1.98s/it, loss=2.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   3%|▎         | 2/79 [00:03<02:28,  1.93s/it, loss=2.76]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   4%|▍         | 3/79 [00:05<02:25,  1.92s/it, loss=2.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   5%|▌         | 4/79 [00:07<02:23,  1.91s/it, loss=2.37]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   6%|▋         | 5/79 [00:09<02:23,  1.94s/it, loss=2.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   8%|▊         | 6/79 [00:11<02:22,  1.95s/it, loss=2.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   9%|▉         | 7/79 [00:13<02:19,  1.94s/it, loss=2.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  10%|█         | 8/79 [00:15<02:17,  1.94s/it, loss=2.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  11%|█▏        | 9/79 [00:17<02:15,  1.94s/it, loss=2.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  13%|█▎        | 10/79 [00:19<02:13,  1.93s/it, loss=2.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  14%|█▍        | 11/79 [00:21<02:11,  1.94s/it, loss=2.44]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  15%|█▌        | 12/79 [00:23<02:11,  1.96s/it, loss=2.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  16%|█▋        | 13/79 [00:25<02:08,  1.95s/it, loss=2.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  18%|█▊        | 14/79 [00:27<02:06,  1.94s/it, loss=2.37]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  19%|█▉        | 15/79 [00:29<02:04,  1.94s/it, loss=2.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  20%|██        | 16/79 [00:31<02:01,  1.94s/it, loss=2.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  22%|██▏       | 17/79 [00:32<01:59,  1.93s/it, loss=2.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  23%|██▎       | 18/79 [00:34<01:59,  1.95s/it, loss=2.3]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  24%|██▍       | 19/79 [00:36<01:57,  1.95s/it, loss=2.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  25%|██▌       | 20/79 [00:38<01:54,  1.94s/it, loss=2.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  27%|██▋       | 21/79 [00:40<01:52,  1.93s/it, loss=2.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  28%|██▊       | 22/79 [00:42<01:49,  1.93s/it, loss=2.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  29%|██▉       | 23/79 [00:44<01:47,  1.92s/it, loss=2.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  30%|███       | 24/79 [00:46<01:45,  1.92s/it, loss=2.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  32%|███▏      | 25/79 [00:48<01:44,  1.94s/it, loss=2.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  33%|███▎      | 26/79 [00:50<01:41,  1.92s/it, loss=2.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  34%|███▍      | 27/79 [00:52<01:39,  1.92s/it, loss=2.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  35%|███▌      | 28/79 [00:54<01:37,  1.91s/it, loss=2.2]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  37%|███▋      | 29/79 [00:56<01:35,  1.91s/it, loss=2.25]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  38%|███▊      | 30/79 [00:57<01:33,  1.90s/it, loss=2.25]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  39%|███▉      | 31/79 [00:59<01:31,  1.91s/it, loss=2.11]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  41%|████      | 32/79 [01:01<01:30,  1.93s/it, loss=2.18]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  42%|████▏     | 33/79 [01:03<01:28,  1.92s/it, loss=2.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  43%|████▎     | 34/79 [01:05<01:26,  1.91s/it, loss=2.16]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  44%|████▍     | 35/79 [01:07<01:23,  1.90s/it, loss=2.19]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  46%|████▌     | 36/79 [01:09<01:21,  1.90s/it, loss=2.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  47%|████▋     | 37/79 [01:11<01:19,  1.90s/it, loss=2.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  48%|████▊     | 38/79 [01:13<01:18,  1.92s/it, loss=2.1]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  49%|████▉     | 39/79 [01:15<01:16,  1.92s/it, loss=2.19]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  51%|█████     | 40/79 [01:17<01:14,  1.91s/it, loss=2.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  52%|█████▏    | 41/79 [01:18<01:12,  1.91s/it, loss=2.04]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  53%|█████▎    | 42/79 [01:20<01:10,  1.90s/it, loss=2.13]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  54%|█████▍    | 43/79 [01:22<01:08,  1.90s/it, loss=2.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  56%|█████▌    | 44/79 [01:24<01:06,  1.91s/it, loss=2.11]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  57%|█████▋    | 45/79 [01:26<01:05,  1.92s/it, loss=2.05]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  58%|█████▊    | 46/79 [01:28<01:03,  1.91s/it, loss=2.3]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  59%|█████▉    | 47/79 [01:30<01:00,  1.91s/it, loss=2.11]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  61%|██████    | 48/79 [01:32<00:58,  1.90s/it, loss=2.1]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  62%|██████▏   | 49/79 [01:34<00:57,  1.90s/it, loss=2.12]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  63%|██████▎   | 50/79 [01:36<00:55,  1.90s/it, loss=2.08]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  65%|██████▍   | 51/79 [01:38<00:53,  1.92s/it, loss=2.04]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  66%|██████▌   | 52/79 [01:40<00:52,  1.94s/it, loss=2.09]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  67%|██████▋   | 53/79 [01:41<00:50,  1.93s/it, loss=2.14]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  68%|██████▊   | 54/79 [01:43<00:48,  1.93s/it, loss=2.11]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  70%|██████▉   | 55/79 [01:45<00:46,  1.92s/it, loss=2]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  71%|███████   | 56/79 [01:47<00:44,  1.92s/it, loss=2.2]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  72%|███████▏  | 57/79 [01:49<00:42,  1.92s/it, loss=2.14]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  73%|███████▎  | 58/79 [01:51<00:40,  1.94s/it, loss=2.13]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  75%|███████▍  | 59/79 [01:53<00:38,  1.93s/it, loss=1.94]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  76%|███████▌  | 60/79 [01:55<00:36,  1.92s/it, loss=1.98]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  77%|███████▋  | 61/79 [01:57<00:34,  1.92s/it, loss=2.14]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  78%|███████▊  | 62/79 [01:59<00:32,  1.92s/it, loss=2.13]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  80%|███████▉  | 63/79 [02:01<00:30,  1.91s/it, loss=1.97]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  81%|████████  | 64/79 [02:03<00:28,  1.92s/it, loss=1.93]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  82%|████████▏ | 65/79 [02:05<00:27,  1.94s/it, loss=2.08]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  84%|████████▎ | 66/79 [02:06<00:25,  1.93s/it, loss=2.01]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  85%|████████▍ | 67/79 [02:08<00:23,  1.93s/it, loss=1.97]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  86%|████████▌ | 68/79 [02:10<00:21,  1.92s/it, loss=1.89]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  87%|████████▋ | 69/79 [02:12<00:19,  1.92s/it, loss=1.91]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  89%|████████▊ | 70/79 [02:14<00:17,  1.92s/it, loss=1.97]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  90%|████████▉ | 71/79 [02:16<00:15,  1.94s/it, loss=1.97]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  91%|█████████ | 72/79 [02:18<00:13,  1.93s/it, loss=1.81]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  92%|█████████▏| 73/79 [02:20<00:11,  1.93s/it, loss=2.07]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  94%|█████████▎| 74/79 [02:22<00:09,  1.92s/it, loss=1.88]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  95%|█████████▍| 75/79 [02:24<00:07,  1.91s/it, loss=1.99]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  96%|█████████▌| 76/79 [02:26<00:05,  1.91s/it, loss=1.89]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  97%|█████████▋| 77/79 [02:28<00:03,  1.92s/it, loss=1.95]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  99%|█████████▊| 78/79 [02:30<00:01,  1.94s/it, loss=1.91]

PGD - Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Mixed Adversarial Training: 100%|██████████| 79/79 [02:30<00:00,  1.90s/it, loss=1.69]


✅ Epoch 1 complete — Avg loss: 2.1707

🌪 Epoch 2/5


Mixed Adversarial Training:   0%|          | 0/79 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   1%|▏         | 1/79 [00:01<02:29,  1.91s/it, loss=1.95]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   3%|▎         | 2/79 [00:03<02:27,  1.91s/it, loss=2.03]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   4%|▍         | 3/79 [00:05<02:25,  1.91s/it, loss=1.98]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   5%|▌         | 4/79 [00:07<02:22,  1.90s/it, loss=1.97]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   6%|▋         | 5/79 [00:09<02:21,  1.92s/it, loss=1.89]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   8%|▊         | 6/79 [00:11<02:21,  1.93s/it, loss=1.91]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   9%|▉         | 7/79 [00:13<02:18,  1.92s/it, loss=1.91]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  10%|█         | 8/79 [00:15<02:15,  1.91s/it, loss=1.97]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  11%|█▏        | 9/79 [00:17<02:13,  1.91s/it, loss=2.04]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  13%|█▎        | 10/79 [00:19<02:11,  1.91s/it, loss=2.02]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  14%|█▍        | 11/79 [00:21<02:09,  1.91s/it, loss=1.89]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  15%|█▌        | 12/79 [00:22<02:08,  1.92s/it, loss=1.92]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  16%|█▋        | 13/79 [00:24<02:07,  1.93s/it, loss=1.86]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  18%|█▊        | 14/79 [00:26<02:05,  1.92s/it, loss=1.94]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  19%|█▉        | 15/79 [00:28<02:02,  1.92s/it, loss=1.94]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  20%|██        | 16/79 [00:30<02:00,  1.91s/it, loss=1.81]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  22%|██▏       | 17/79 [00:32<01:58,  1.91s/it, loss=1.87]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  23%|██▎       | 18/79 [00:34<01:56,  1.91s/it, loss=1.96]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  24%|██▍       | 19/79 [00:36<01:55,  1.93s/it, loss=1.88]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  25%|██▌       | 20/79 [00:38<01:53,  1.93s/it, loss=1.94]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  27%|██▋       | 21/79 [00:40<01:51,  1.92s/it, loss=1.84]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  28%|██▊       | 22/79 [00:42<01:49,  1.92s/it, loss=1.83]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  29%|██▉       | 23/79 [00:44<01:47,  1.91s/it, loss=1.95]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  30%|███       | 24/79 [00:45<01:45,  1.91s/it, loss=1.75]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  32%|███▏      | 25/79 [00:47<01:43,  1.92s/it, loss=1.82]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  33%|███▎      | 26/79 [00:49<01:42,  1.93s/it, loss=1.86]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  34%|███▍      | 27/79 [00:51<01:40,  1.93s/it, loss=1.68]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  35%|███▌      | 28/79 [00:53<01:38,  1.92s/it, loss=1.77]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  37%|███▋      | 29/79 [00:55<01:35,  1.91s/it, loss=1.75]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  38%|███▊      | 30/79 [00:57<01:33,  1.91s/it, loss=1.65]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  39%|███▉      | 31/79 [00:59<01:31,  1.90s/it, loss=1.63]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  41%|████      | 32/79 [01:01<01:30,  1.92s/it, loss=1.83]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  42%|████▏     | 33/79 [01:03<01:28,  1.93s/it, loss=1.78]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  43%|████▎     | 34/79 [01:05<01:27,  1.93s/it, loss=1.91]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  44%|████▍     | 35/79 [01:07<01:24,  1.93s/it, loss=1.99]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  46%|████▌     | 36/79 [01:09<01:22,  1.92s/it, loss=1.87]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  47%|████▋     | 37/79 [01:11<01:20,  1.92s/it, loss=1.77]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  48%|████▊     | 38/79 [01:12<01:18,  1.92s/it, loss=1.67]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  49%|████▉     | 39/79 [01:14<01:17,  1.94s/it, loss=1.73]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  51%|█████     | 40/79 [01:16<01:15,  1.93s/it, loss=1.74]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  52%|█████▏    | 41/79 [01:18<01:13,  1.93s/it, loss=1.78]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  53%|█████▎    | 42/79 [01:20<01:11,  1.92s/it, loss=1.73]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  54%|█████▍    | 43/79 [01:22<01:08,  1.92s/it, loss=1.67]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  56%|█████▌    | 44/79 [01:24<01:06,  1.91s/it, loss=1.88]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  57%|█████▋    | 45/79 [01:26<01:05,  1.93s/it, loss=2.04]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  58%|█████▊    | 46/79 [01:28<01:04,  1.95s/it, loss=1.73]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  59%|█████▉    | 47/79 [01:30<01:01,  1.93s/it, loss=1.88]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  61%|██████    | 48/79 [01:32<00:59,  1.92s/it, loss=1.93]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  62%|██████▏   | 49/79 [01:34<00:57,  1.92s/it, loss=1.83]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  63%|██████▎   | 50/79 [01:36<00:55,  1.91s/it, loss=1.73]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  65%|██████▍   | 51/79 [01:37<00:53,  1.92s/it, loss=1.67]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  66%|██████▌   | 52/79 [01:39<00:52,  1.94s/it, loss=1.71]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  67%|██████▋   | 53/79 [01:41<00:50,  1.93s/it, loss=1.64]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  68%|██████▊   | 54/79 [01:43<00:48,  1.93s/it, loss=1.67]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  70%|██████▉   | 55/79 [01:45<00:46,  1.92s/it, loss=1.74]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  71%|███████   | 56/79 [01:47<00:44,  1.92s/it, loss=1.61]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  72%|███████▏  | 57/79 [01:49<00:42,  1.91s/it, loss=1.7]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  73%|███████▎  | 58/79 [01:51<00:40,  1.92s/it, loss=1.81]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  75%|███████▍  | 59/79 [01:53<00:38,  1.94s/it, loss=1.52]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  76%|███████▌  | 60/79 [01:55<00:36,  1.93s/it, loss=1.64]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  77%|███████▋  | 61/79 [01:57<00:34,  1.92s/it, loss=1.71]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  78%|███████▊  | 62/79 [01:59<00:32,  1.92s/it, loss=1.57]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  80%|███████▉  | 63/79 [02:01<00:30,  1.91s/it, loss=1.64]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  81%|████████  | 64/79 [02:02<00:28,  1.91s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  82%|████████▏ | 65/79 [02:04<00:27,  1.93s/it, loss=1.54]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  84%|████████▎ | 66/79 [02:06<00:25,  1.94s/it, loss=1.58]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  85%|████████▍ | 67/79 [02:08<00:23,  1.93s/it, loss=1.59]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  86%|████████▌ | 68/79 [02:10<00:21,  1.92s/it, loss=1.72]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  87%|████████▋ | 69/79 [02:12<00:19,  1.92s/it, loss=1.8]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  89%|████████▊ | 70/79 [02:14<00:17,  1.92s/it, loss=1.56]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  90%|████████▉ | 71/79 [02:16<00:15,  1.93s/it, loss=1.57]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  91%|█████████ | 72/79 [02:18<00:13,  1.94s/it, loss=1.69]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  92%|█████████▏| 73/79 [02:20<00:11,  1.93s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  94%|█████████▎| 74/79 [02:22<00:09,  1.92s/it, loss=1.55]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  95%|█████████▍| 75/79 [02:24<00:07,  1.92s/it, loss=1.71]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  96%|█████████▌| 76/79 [02:26<00:05,  1.92s/it, loss=1.59]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  97%|█████████▋| 77/79 [02:27<00:03,  1.91s/it, loss=1.64]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  99%|█████████▊| 78/79 [02:29<00:01,  1.94s/it, loss=1.66]

PGD - Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Mixed Adversarial Training: 100%|██████████| 79/79 [02:30<00:00,  1.90s/it, loss=1.64]


✅ Epoch 2 complete — Avg loss: 1.7740

🌪 Epoch 3/5


Mixed Adversarial Training:   0%|          | 0/79 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   1%|▏         | 1/79 [00:01<02:33,  1.97s/it, loss=1.63]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   3%|▎         | 2/79 [00:03<02:28,  1.93s/it, loss=1.59]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   4%|▍         | 3/79 [00:05<02:26,  1.92s/it, loss=1.59]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   5%|▌         | 4/79 [00:07<02:23,  1.92s/it, loss=1.62]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   6%|▋         | 5/79 [00:09<02:21,  1.91s/it, loss=1.55]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   8%|▊         | 6/79 [00:11<02:20,  1.92s/it, loss=1.55]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   9%|▉         | 7/79 [00:13<02:19,  1.94s/it, loss=1.53]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  10%|█         | 8/79 [00:15<02:17,  1.93s/it, loss=1.55]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  11%|█▏        | 9/79 [00:17<02:14,  1.93s/it, loss=1.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  13%|█▎        | 10/79 [00:19<02:12,  1.92s/it, loss=1.58]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  14%|█▍        | 11/79 [00:21<02:10,  1.91s/it, loss=1.56]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  15%|█▌        | 12/79 [00:23<02:08,  1.91s/it, loss=1.45]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  16%|█▋        | 13/79 [00:25<02:07,  1.93s/it, loss=1.61]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  18%|█▊        | 14/79 [00:27<02:06,  1.95s/it, loss=1.68]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  19%|█▉        | 15/79 [00:28<02:03,  1.94s/it, loss=1.59]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  20%|██        | 16/79 [00:30<02:01,  1.93s/it, loss=1.45]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  22%|██▏       | 17/79 [00:32<01:59,  1.92s/it, loss=1.41]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  23%|██▎       | 18/79 [00:34<01:56,  1.92s/it, loss=1.49]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  24%|██▍       | 19/79 [00:36<01:54,  1.92s/it, loss=1.46]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  25%|██▌       | 20/79 [00:38<01:54,  1.93s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  27%|██▋       | 21/79 [00:40<01:52,  1.94s/it, loss=1.58]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  28%|██▊       | 22/79 [00:42<01:50,  1.93s/it, loss=1.75]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  29%|██▉       | 23/79 [00:44<01:47,  1.92s/it, loss=1.41]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  30%|███       | 24/79 [00:46<01:45,  1.92s/it, loss=1.45]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  32%|███▏      | 25/79 [00:48<01:43,  1.92s/it, loss=1.46]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  33%|███▎      | 26/79 [00:50<01:41,  1.92s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  34%|███▍      | 27/79 [00:52<01:40,  1.94s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  35%|███▌      | 28/79 [00:53<01:38,  1.93s/it, loss=1.6]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  37%|███▋      | 29/79 [00:55<01:36,  1.92s/it, loss=1.4]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  38%|███▊      | 30/79 [00:58<01:38,  2.00s/it, loss=1.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  39%|███▉      | 31/79 [00:59<01:34,  1.97s/it, loss=1.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  41%|████      | 32/79 [01:01<01:31,  1.95s/it, loss=1.63]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  42%|████▏     | 33/79 [01:03<01:30,  1.96s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  43%|████▎     | 34/79 [01:05<01:28,  1.97s/it, loss=1.54]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  44%|████▍     | 35/79 [01:07<01:25,  1.95s/it, loss=1.57]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  46%|████▌     | 36/79 [01:09<01:23,  1.94s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  47%|████▋     | 37/79 [01:11<01:20,  1.93s/it, loss=1.52]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  48%|████▊     | 38/79 [01:13<01:18,  1.92s/it, loss=1.54]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  49%|████▉     | 39/79 [01:15<01:16,  1.92s/it, loss=1.4]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  51%|█████     | 40/79 [01:17<01:15,  1.94s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  52%|█████▏    | 41/79 [01:19<01:13,  1.94s/it, loss=1.37]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  53%|█████▎    | 42/79 [01:21<01:11,  1.92s/it, loss=1.49]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  54%|█████▍    | 43/79 [01:23<01:09,  1.92s/it, loss=1.55]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  56%|█████▌    | 44/79 [01:24<01:06,  1.91s/it, loss=1.4]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  57%|█████▋    | 45/79 [01:26<01:05,  1.91s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  58%|█████▊    | 46/79 [01:28<01:03,  1.92s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  59%|█████▉    | 47/79 [01:30<01:01,  1.94s/it, loss=1.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  61%|██████    | 48/79 [01:32<00:59,  1.93s/it, loss=1.57]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  62%|██████▏   | 49/79 [01:34<00:57,  1.92s/it, loss=1.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  63%|██████▎   | 50/79 [01:36<00:55,  1.92s/it, loss=1.39]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  65%|██████▍   | 51/79 [01:38<00:53,  1.92s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  66%|██████▌   | 52/79 [01:40<00:51,  1.92s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  67%|██████▋   | 53/79 [01:42<00:50,  1.94s/it, loss=1.55]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  68%|██████▊   | 54/79 [01:44<00:48,  1.94s/it, loss=1.45]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  70%|██████▉   | 55/79 [01:46<00:46,  1.93s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  71%|███████   | 56/79 [01:48<00:44,  1.92s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  72%|███████▏  | 57/79 [01:50<00:42,  1.92s/it, loss=1.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  73%|███████▎  | 58/79 [01:51<00:40,  1.92s/it, loss=1.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  75%|███████▍  | 59/79 [01:53<00:38,  1.93s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  76%|███████▌  | 60/79 [01:55<00:36,  1.94s/it, loss=1.4]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  77%|███████▋  | 61/79 [01:57<00:34,  1.93s/it, loss=1.52]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  78%|███████▊  | 62/79 [01:59<00:32,  1.93s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  80%|███████▉  | 63/79 [02:01<00:30,  1.92s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  81%|████████  | 64/79 [02:03<00:28,  1.92s/it, loss=1.46]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  82%|████████▏ | 65/79 [02:05<00:26,  1.92s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  84%|████████▎ | 66/79 [02:07<00:25,  1.94s/it, loss=1.47]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  85%|████████▍ | 67/79 [02:09<00:23,  1.94s/it, loss=1.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  86%|████████▌ | 68/79 [02:11<00:21,  1.93s/it, loss=1.39]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  87%|████████▋ | 69/79 [02:13<00:19,  1.92s/it, loss=1.59]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  89%|████████▊ | 70/79 [02:15<00:17,  1.92s/it, loss=1.43]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  90%|████████▉ | 71/79 [02:17<00:15,  1.92s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  91%|█████████ | 72/79 [02:18<00:13,  1.93s/it, loss=1.45]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  92%|█████████▏| 73/79 [02:20<00:11,  1.95s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  94%|█████████▎| 74/79 [02:22<00:09,  1.93s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  95%|█████████▍| 75/79 [02:24<00:07,  1.93s/it, loss=1.26]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  96%|█████████▌| 76/79 [02:26<00:05,  1.92s/it, loss=1.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  97%|█████████▋| 77/79 [02:28<00:03,  1.92s/it, loss=1.4]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  99%|█████████▊| 78/79 [02:30<00:01,  1.92s/it, loss=1.26]

PGD - Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Mixed Adversarial Training: 100%|██████████| 79/79 [02:30<00:00,  1.91s/it, loss=1.57]


✅ Epoch 3 complete — Avg loss: 1.4645

🌪 Epoch 4/5


Mixed Adversarial Training:   0%|          | 0/79 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   1%|▏         | 1/79 [00:02<02:36,  2.00s/it, loss=1.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   3%|▎         | 2/79 [00:03<02:30,  1.95s/it, loss=1.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   4%|▍         | 3/79 [00:05<02:27,  1.94s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   5%|▌         | 4/79 [00:07<02:24,  1.92s/it, loss=1.52]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   6%|▋         | 5/79 [00:09<02:22,  1.92s/it, loss=1.41]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   8%|▊         | 6/79 [00:11<02:19,  1.92s/it, loss=1.43]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   9%|▉         | 7/79 [00:13<02:20,  1.95s/it, loss=1.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  10%|█         | 8/79 [00:15<02:19,  1.96s/it, loss=1.46]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  11%|█▏        | 9/79 [00:17<02:16,  1.95s/it, loss=1.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  13%|█▎        | 10/79 [00:19<02:13,  1.94s/it, loss=1.52]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  14%|█▍        | 11/79 [00:21<02:11,  1.93s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  15%|█▌        | 12/79 [00:23<02:08,  1.92s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  16%|█▋        | 13/79 [00:25<02:07,  1.93s/it, loss=1.41]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  18%|█▊        | 14/79 [00:27<02:06,  1.95s/it, loss=1.49]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  19%|█▉        | 15/79 [00:29<02:03,  1.94s/it, loss=1.37]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  20%|██        | 16/79 [00:30<02:01,  1.93s/it, loss=1.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  22%|██▏       | 17/79 [00:32<01:59,  1.92s/it, loss=1.41]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  23%|██▎       | 18/79 [00:34<01:56,  1.92s/it, loss=1.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  24%|██▍       | 19/79 [00:36<01:55,  1.92s/it, loss=1.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  25%|██▌       | 20/79 [00:38<01:54,  1.94s/it, loss=1.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  27%|██▋       | 21/79 [00:40<01:53,  1.95s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  28%|██▊       | 22/79 [00:42<01:50,  1.95s/it, loss=1.44]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  29%|██▉       | 23/79 [00:44<01:48,  1.94s/it, loss=1.48]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  30%|███       | 24/79 [00:46<01:46,  1.93s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  32%|███▏      | 25/79 [00:48<01:43,  1.92s/it, loss=1.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  33%|███▎      | 26/79 [00:50<01:41,  1.92s/it, loss=1.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  34%|███▍      | 27/79 [00:52<01:41,  1.95s/it, loss=1.44]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  35%|███▌      | 28/79 [00:54<01:38,  1.94s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  37%|███▋      | 29/79 [00:56<01:36,  1.93s/it, loss=1.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  38%|███▊      | 30/79 [00:58<01:34,  1.93s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  39%|███▉      | 31/79 [00:59<01:32,  1.92s/it, loss=1.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  41%|████      | 32/79 [01:01<01:30,  1.92s/it, loss=1.45]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  42%|████▏     | 33/79 [01:03<01:29,  1.94s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  43%|████▎     | 34/79 [01:05<01:27,  1.95s/it, loss=1.26]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  44%|████▍     | 35/79 [01:07<01:25,  1.94s/it, loss=1.67]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  46%|████▌     | 36/79 [01:09<01:23,  1.93s/it, loss=1.51]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  47%|████▋     | 37/79 [01:11<01:20,  1.93s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  48%|████▊     | 38/79 [01:13<01:18,  1.92s/it, loss=1.42]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  49%|████▉     | 39/79 [01:15<01:17,  1.93s/it, loss=1.18]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  51%|█████     | 40/79 [01:17<01:16,  1.95s/it, loss=1.29]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  52%|█████▏    | 41/79 [01:19<01:13,  1.94s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  53%|█████▎    | 42/79 [01:21<01:11,  1.93s/it, loss=1.29]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  54%|█████▍    | 43/79 [01:23<01:09,  1.92s/it, loss=1.41]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  56%|█████▌    | 44/79 [01:25<01:07,  1.92s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  57%|█████▋    | 45/79 [01:26<01:05,  1.92s/it, loss=1.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  58%|█████▊    | 46/79 [01:28<01:03,  1.93s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  59%|█████▉    | 47/79 [01:30<01:02,  1.94s/it, loss=1.19]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  61%|██████    | 48/79 [01:32<01:00,  1.94s/it, loss=1.17]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  62%|██████▏   | 49/79 [01:34<00:57,  1.93s/it, loss=1.2]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  63%|██████▎   | 50/79 [01:36<00:55,  1.92s/it, loss=1.24]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  65%|██████▍   | 51/79 [01:38<00:53,  1.92s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  66%|██████▌   | 52/79 [01:40<00:52,  1.93s/it, loss=1.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  67%|██████▋   | 53/79 [01:42<00:50,  1.95s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  68%|██████▊   | 54/79 [01:44<00:48,  1.94s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  70%|██████▉   | 55/79 [01:46<00:46,  1.93s/it, loss=1.46]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  71%|███████   | 56/79 [01:48<00:44,  1.92s/it, loss=1.24]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  72%|███████▏  | 57/79 [01:50<00:42,  1.92s/it, loss=1.39]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  73%|███████▎  | 58/79 [01:52<00:40,  1.92s/it, loss=1.16]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  75%|███████▍  | 59/79 [01:54<00:38,  1.93s/it, loss=1.44]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  76%|███████▌  | 60/79 [01:55<00:36,  1.95s/it, loss=1.25]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  77%|███████▋  | 61/79 [01:57<00:34,  1.93s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  78%|███████▊  | 62/79 [01:59<00:32,  1.92s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  80%|███████▉  | 63/79 [02:01<00:30,  1.92s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  81%|████████  | 64/79 [02:03<00:28,  1.92s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  82%|████████▏ | 65/79 [02:05<00:26,  1.92s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  84%|████████▎ | 66/79 [02:07<00:25,  1.94s/it, loss=1.64]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  85%|████████▍ | 67/79 [02:09<00:23,  1.94s/it, loss=1.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  86%|████████▌ | 68/79 [02:11<00:21,  1.93s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  87%|████████▋ | 69/79 [02:13<00:19,  1.92s/it, loss=1.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  89%|████████▊ | 70/79 [02:15<00:17,  1.92s/it, loss=1.25]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  90%|████████▉ | 71/79 [02:17<00:15,  1.92s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  91%|█████████ | 72/79 [02:19<00:13,  1.93s/it, loss=1.29]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  92%|█████████▏| 73/79 [02:21<00:11,  1.94s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  94%|█████████▎| 74/79 [02:22<00:09,  1.93s/it, loss=1.26]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  95%|█████████▍| 75/79 [02:24<00:07,  1.93s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  96%|█████████▌| 76/79 [02:26<00:05,  1.93s/it, loss=1.16]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  97%|█████████▋| 77/79 [02:28<00:03,  1.92s/it, loss=1.04]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  99%|█████████▊| 78/79 [02:30<00:01,  1.93s/it, loss=1.23]

PGD - Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Mixed Adversarial Training: 100%|██████████| 79/79 [02:31<00:00,  1.91s/it, loss=1.32]


✅ Epoch 4 complete — Avg loss: 1.3361

🌪 Epoch 5/5


Mixed Adversarial Training:   0%|          | 0/79 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   1%|▏         | 1/79 [00:01<02:35,  1.99s/it, loss=1.2]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   3%|▎         | 2/79 [00:03<02:30,  1.95s/it, loss=1.3]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   4%|▍         | 3/79 [00:05<02:27,  1.94s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   5%|▌         | 4/79 [00:07<02:24,  1.93s/it, loss=1.08]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   6%|▋         | 5/79 [00:09<02:21,  1.92s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   8%|▊         | 6/79 [00:11<02:19,  1.91s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:   9%|▉         | 7/79 [00:13<02:18,  1.93s/it, loss=1.3]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  10%|█         | 8/79 [00:15<02:18,  1.95s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  11%|█▏        | 9/79 [00:17<02:15,  1.94s/it, loss=1.17]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  13%|█▎        | 10/79 [00:19<02:12,  1.93s/it, loss=1.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  14%|█▍        | 11/79 [00:21<02:10,  1.92s/it, loss=1.29]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  15%|█▌        | 12/79 [00:23<02:08,  1.91s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  16%|█▋        | 13/79 [00:25<02:06,  1.92s/it, loss=1.35]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  18%|█▊        | 14/79 [00:27<02:06,  1.94s/it, loss=1.36]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  19%|█▉        | 15/79 [00:28<02:03,  1.93s/it, loss=1.46]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  20%|██        | 16/79 [00:30<02:01,  1.93s/it, loss=1.16]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  22%|██▏       | 17/79 [00:32<01:59,  1.92s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  23%|██▎       | 18/79 [00:34<01:57,  1.92s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  24%|██▍       | 19/79 [00:36<01:55,  1.92s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  25%|██▌       | 20/79 [00:38<01:54,  1.93s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  27%|██▋       | 21/79 [00:40<01:53,  1.95s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  28%|██▊       | 22/79 [00:42<01:50,  1.94s/it, loss=1.24]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  29%|██▉       | 23/79 [00:44<01:48,  1.93s/it, loss=1.4]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  30%|███       | 24/79 [00:46<01:45,  1.93s/it, loss=1.18]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  32%|███▏      | 25/79 [00:48<01:43,  1.92s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  33%|███▎      | 26/79 [00:50<01:41,  1.92s/it, loss=1.26]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  34%|███▍      | 27/79 [00:52<01:41,  1.94s/it, loss=1.17]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  35%|███▌      | 28/79 [00:54<01:38,  1.93s/it, loss=1.32]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  37%|███▋      | 29/79 [00:55<01:36,  1.93s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  38%|███▊      | 30/79 [00:57<01:34,  1.92s/it, loss=1.1]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  39%|███▉      | 31/79 [00:59<01:31,  1.92s/it, loss=1.09]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  41%|████      | 32/79 [01:01<01:29,  1.91s/it, loss=1.07]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  42%|████▏     | 33/79 [01:03<01:28,  1.93s/it, loss=1.15]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  43%|████▎     | 34/79 [01:05<01:27,  1.95s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  44%|████▍     | 35/79 [01:07<01:25,  1.94s/it, loss=1.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  46%|████▌     | 36/79 [01:09<01:23,  1.93s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  47%|████▋     | 37/79 [01:11<01:21,  1.93s/it, loss=1.16]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  48%|████▊     | 38/79 [01:13<01:19,  1.94s/it, loss=1.33]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  49%|████▉     | 39/79 [01:15<01:17,  1.94s/it, loss=1.28]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  51%|█████     | 40/79 [01:17<01:16,  1.96s/it, loss=1.1]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  52%|█████▏    | 41/79 [01:19<01:14,  1.95s/it, loss=1.1]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  53%|█████▎    | 42/79 [01:21<01:11,  1.94s/it, loss=1.12]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  54%|█████▍    | 43/79 [01:23<01:09,  1.93s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  56%|█████▌    | 44/79 [01:24<01:07,  1.92s/it, loss=1.2]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  57%|█████▋    | 45/79 [01:26<01:05,  1.92s/it, loss=1.11]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  58%|█████▊    | 46/79 [01:28<01:03,  1.94s/it, loss=1.18]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  59%|█████▉    | 47/79 [01:30<01:02,  1.95s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  61%|██████    | 48/79 [01:32<01:00,  1.94s/it, loss=1.19]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  62%|██████▏   | 49/79 [01:34<00:57,  1.93s/it, loss=1.18]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  63%|██████▎   | 50/79 [01:36<00:55,  1.93s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  65%|██████▍   | 51/79 [01:38<00:53,  1.93s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  66%|██████▌   | 52/79 [01:40<00:52,  1.93s/it, loss=1.11]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  67%|██████▋   | 53/79 [01:42<00:50,  1.96s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  68%|██████▊   | 54/79 [01:44<00:48,  1.95s/it, loss=1.27]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  70%|██████▉   | 55/79 [01:46<00:46,  1.94s/it, loss=1.25]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  71%|███████   | 56/79 [01:48<00:44,  1.93s/it, loss=1.34]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  72%|███████▏  | 57/79 [01:50<00:42,  1.92s/it, loss=1.26]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  73%|███████▎  | 58/79 [01:52<00:40,  1.92s/it, loss=1.12]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  75%|███████▍  | 59/79 [01:54<00:38,  1.93s/it, loss=1.03]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  76%|███████▌  | 60/79 [01:56<00:37,  1.95s/it, loss=1.31]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  77%|███████▋  | 61/79 [01:57<00:34,  1.93s/it, loss=1.38]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  78%|███████▊  | 62/79 [01:59<00:32,  1.93s/it, loss=1.18]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  80%|███████▉  | 63/79 [02:01<00:30,  1.92s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  81%|████████  | 64/79 [02:03<00:28,  1.92s/it, loss=1.09]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  82%|████████▏ | 65/79 [02:05<00:26,  1.92s/it, loss=1.14]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  84%|████████▎ | 66/79 [02:07<00:25,  1.95s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  85%|████████▍ | 67/79 [02:09<00:23,  1.94s/it, loss=1.29]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  86%|████████▌ | 68/79 [02:11<00:21,  1.93s/it, loss=1.16]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  87%|████████▋ | 69/79 [02:13<00:19,  1.93s/it, loss=1.15]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  89%|████████▊ | 70/79 [02:15<00:17,  1.92s/it, loss=1.22]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  90%|████████▉ | 71/79 [02:17<00:15,  1.92s/it, loss=1.07]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  91%|█████████ | 72/79 [02:19<00:13,  1.93s/it, loss=1.21]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  92%|█████████▏| 73/79 [02:21<00:11,  1.94s/it, loss=1.17]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  94%|█████████▎| 74/79 [02:22<00:09,  1.93s/it, loss=1.26]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  95%|█████████▍| 75/79 [02:24<00:07,  1.92s/it, loss=1.23]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  96%|█████████▌| 76/79 [02:26<00:05,  1.92s/it, loss=1.37]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  97%|█████████▋| 77/79 [02:28<00:03,  1.93s/it, loss=1.12]

PGD - Batches:   0%|          | 0/4 [00:00<?, ?it/s]

Mixed Adversarial Training:  99%|█████████▊| 78/79 [02:30<00:01,  1.92s/it, loss=1.22]

PGD - Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Mixed Adversarial Training: 100%|██████████| 79/79 [02:31<00:00,  1.91s/it, loss=1.53]

✅ Epoch 5 complete — Avg loss: 1.2303


Accuracy evaluation on clean + FGSM + PGD:

In [ ]:
from art.attacks.evasion import FastGradientMethod, ProjectedGradientDescent
from art.estimators.classification import PyTorchClassifier
import torch.nn.functional as F

# Убедимся, что модель в режиме инференса
model.eval()

# 1. Точность на чистых данных
correct = 0
total = 0

with torch.no_grad():
    for images, labels in test_loader:
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)
        preds = outputs.argmax(dim=1)
        correct += (preds == labels).sum().item()
        total += labels.size(0)

acc_clean = 100.0 * correct / total
print(f"🎯 Accuracy on clean test set: {acc_clean:.2f}%")

# 2. Генерация adversarial атак (по батчам)
def evaluate_adversarial(attack):
    correct = 0
    total = 0
    model.eval()

    for images, labels in test_loader:
        images, labels = images.to(device), labels.to(device)

        # Переводим в numpy и атакуем
        adv_images = attack.generate(x=images.detach().cpu().numpy())

        # Переводим обратно в тензор
        adv_images = torch.tensor(adv_images).to(device).float()

        with torch.no_grad():
            outputs = model(adv_images)
            preds = outputs.argmax(dim=1)
            correct += (preds == labels).sum().item()
            total += labels.size(0)

    return 100.0 * correct / total

# 3. FGSM
fgsm = FastGradientMethod(estimator=classifier, eps=0.03)
acc_fgsm = evaluate_adversarial(fgsm)
print(f"⚠️ Accuracy under FGSM attack: {acc_fgsm:.2f}%")

# 4. PGD
pgd = ProjectedGradientDescent(estimator=classifier, eps=0.03, eps_step=0.007, max_iter=40)
acc_pgd = evaluate_adversarial(pgd)
print(f"🔥 Accuracy under PGD attack: {acc_pgd:.2f}%")


🎯 Accuracy on clean test set: 75.06%
⚠️ Accuracy under FGSM attack: 41.10%


PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/8 [00:00<?, ?it/s]

PGD - Batches:   0%|          | 0/6 [00:00<?, ?it/s]

🔥 Accuracy under PGD attack: 31.80%


From the results, we can see that the accuracy has decreased on both normal and attacked images. This may be due to the fact that I trained on a subset of 79 values (instead of the full dataset as in the previous training), with a batch size of 256 (instead of 128). However, training would have taken approximately 5 hours if I had used the full dataset.

It is also worth considering the following factors:

During adversarial training, the model learns to be more resistant to attacks, but at the same time, it loses some of its accuracy on regular data.
In other words, improving resistance often comes at the cost of reducing pure accuracy.

**The complexity of mixed adversarial training**
Training on two types of attacks (FGSM and PGD) simultaneously is more challenging and may require more careful selection of hyperparameters (learning rate, epochs, batch size). Otherwise, the model may get confused or retrain on simpler examples.

The task does not specify whether the model should be able to adapt to one or multiple attacks. Therefore, in this case, the resnet18 model, after being trained on a modified dataset, became more resistant to the PGD attack (with an accuracy that increased from 6.59% to 54.61%).